synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -237,7 +240,12 @@ async def action_for_event_by_user(self, event, context) -> None:
             # The push rules endpoint on the CS API checks the actions on new push rules
             # and limit them to spec'd ones, so we shouldn't have to worry about users
             # changing their push rules to include this action.
-            if _should_count_as_unread(event, context):
+            # An alternative way to do this would be to pass count_as_unread directly to
+            # add_push_actions_to_staging, but that wouldn't work as that function relies
+            # on actions_by_user to determine the list of users to insert a row for,
+            # therefore it wouldn't add a row for any user that's in the room but doesn't
+            # get notified by the event.
```

Comment:
we could ensure that `actions_by_user` has an entry for each user? (initialise it with `actions_by_user[uid] = []` before the `for rule in rules:` list)

using `mark_unread` as a "magic" action seems a bit unintuitive to me.

Reply:
Oh yeah that works too. Not sure why I didn't try this in the first place.